Commit for 1st code review

simeo-potiron opened the pull request:
Template PR pas encore prêt.
Juste prêt pour la 1ère review.

simeo-potiron committed:
Commit for 1st code review

simeo-potiron committed:
Finishing preparing 1st merge

Add all necessary files
Few code modifications

simeo-potiron committed:
Correction workflow test

simeo-potiron committed:
Test new change on run tests

simeo-potiron committed:
Final tests config

simeo-potiron committed:
Add template test

LouisMonierSpark commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,20 @@
+### Summary
```
Ce template doit être adapté si il est voué à apparaitre lors d'une demande de PR de la part d'une personne hors de l'organization.

simeo-potiron replied:
Ah yes, tu voudrais ajouter/enlever quoi ?

LouisMonierSpark replied:
# Pull Request Title

## Description
<!-- Provide a concise description of the changes introduced by this PR. -->

## Related Issue
<!-- Link to the issue this PR addresses (e.g., closes #123). -->
Closes #IssueNumber

## Type of Change
<!-- Please delete options that are not relevant. -->
- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update
- [ ] Refactoring
- [ ] Style (formatting, etc.)
- [ ] Other (please specify):

## How Has This Been Tested?
<!-- Describe the tests you've run to verify your changes. -->
- Test 1
- Test 2
- Test 3

## Screenshots (if applicable)
<!-- Add screenshots to help explain your changes. -->
![Screenshot Description](URL)

## Checklist
- [ ] My code follows the project's coding style.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added necessary documentation (if applicable).
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] All new and existing tests pass locally with my changes.

## Additional Information
<!-- Any additional context or information related to this PR. -->

LouisMonierSpark commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,11 @@
+# Summary
```
+1 pour l'idée. 
Tu peux t'inspirer de ça par exemple pour que les reports soient plus complet: https://github.com/metabase/metabase/blob/master/.github/ISSUE_TEMPLATE/bug_report.yml

simeo-potiron replied:
J'ai copié/collé (en adaptant) leur système d'issue template

LouisMonierSpark commented on .github/workflows/publish_new_version.yml:
```diff
@@ -0,0 +1,35 @@
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.8"
```
Pourquoi ne pas utiliser une version de python plus actuelle ?

simeo-potiron replied:
Indeed, coquille
Je mets 3.13, mais j'ai checké et il n'est même pas forcément nécessaire de préciser de version de python

LouisMonierSpark commented on .github/workflows/test_new_pull_request.yml:
```diff
@@ -0,0 +1,31 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [ "pypy3.10", "3.9", "3.10", "3.11", "3.12", "3.13" ]
```
Je me suis renseigné un peu et en effet, ce sont les versions python encore maintenues à ce jour. pypy3.10 je connaissais pas: est-ce qu'on en a vraiment besoin dans ce projet ? Ca semble être un truc de niche.

simeo-potiron replied:
Ok, j'enlève

LouisMonierSpark commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Tous les fichiers dans ce dossier _.idea_ sont propres à la configuration locale. Ils peuvent être retirés du repo.

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+media networks and search engines, disclose their advertising practices, including providing access to data on the
+targeting of ads and the related algorithms.
+
+The purpose of the DSA is to combat harmful content, enhance user safety, and ensure that platforms disclose details
```
C'est toi qui a rédigé ça ?

simeo-potiron replied:
Oui, pas pertinent selon toi ?

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+
+[Learn more about the Digital Services Act (DSA)](https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A32022R2065)
+
+The Nanga Ad Library SDK allows you to query the ad library of several social media platforms.
```
Mettre une roadmap pour indiquer ce qui est dispo (comme plateformes) et ce qui ne l'est pas.

simeo-potiron replied:
Done, bien vu

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+[Learn more about the Digital Services Act (DSA)](https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A32022R2065)
+
+The Nanga Ad Library SDK allows you to query the ad library of several social media platforms.
+You only need to provide access tokens for each platform, and you'll be able to retrieve rich data about
```
"." final

simeo-potiron replied:
Oups, oublié de rédiger la fin...

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+The regulations are enforced by the European Commission, and platforms that fail to comply face significant penalties,
+including fines of up to 6% of their global revenue.
+
+[Learn more about the Digital Services Act (DSA)](https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A32022R2065)
```
"." final

simeo-potiron replied:
Ok

LouisMonierSpark commented on README.md:
````diff
@@ -1,2 +1,126 @@
+
+You can install this package directly from PyPI using `pip`:
+```bash
+pip install nanga_ad_library
````
Peux-tu changer les "_" pour "-" dans le nom du package lui-même ?

simeo-potiron replied:
yes

LouisMonierSpark commented on .idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Supprimer tout le dossier _**.idea**_

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+
+This command will automatically download and install all required dependencies.  
+
+### Updating
```
Ca me semble évident.

simeo-potiron replied:
Ok, enlevé

LouisMonierSpark commented on README.md:
````diff
@@ -1,2 +1,126 @@
+pip install --upgrade nanga_ad_library
+```
+
+### Verifying the Installation
````
Ca me semble évident "2".

simeo-potiron replied:
Ok

LouisMonierSpark commented on README.md:
````diff
@@ -1,2 +1,126 @@
+
+Try to extract some results from the Nanga Ad Library API:
+```python
+from nanga_ad_library import NangaAdLibraryApi
````
"-" à la place de "_" serait le top

simeo-potiron replied:
Pour moi, '-' dans le projet PyPI, mais '_' dans le module python (cf python naming convention)

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+# Prepare query hash (here is an example for Meta):
+query_hash = {
+    "payload": {
+        "search_terms": "Facebook",
```
1 par 1 ? 
Peut-on passer une liste ?

simeo-potiron replied:
Non, il faut séparer les mots clés par des espaces

LouisMonierSpark replied:
Est-ce que c'est indiqué/marqué quelque part ? 
Si j'arrive sur le repo pour la 1ère fois, comment je peux le savoir ?

simeo-potiron replied:
C'est dans la doc Meta, mais je vais l'ajouter à la doc du répo

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+        "search_terms": "Facebook",
+        "ad_reached_countries": ["FR"]
+    },
+    "fields": []
```
Pas de fields à préciser ?

simeo-potiron replied:
Il y a des champs par défaut (cf les classes Iter), on peut en préciser d'autres lorsque nécessaires.

LouisMonierSpark replied:
Est-ce qu'on peut avoir la liste précise des champs par défaut/optionnels dans la doc ?

simeo-potiron replied:
Yes il faut que j'ajoute une façon de le rendre disponible.
Tu aurais une idée ? 
Je me disais pourquoi pas avoir une fonction intégrée (nommée help par ex) qui prendrait en param la plateforme et pourrait renvoyer un payload template par exemple et afficher quelques infos utiles (les params nécessaires, les champs dispos, etc...)
Peu être plus dans un 2ème temps comme ce n'est pas nécessaire pour l'utilisation en interne.
Ce serait l'occas de créer la 1ère issue du répo, wdyt ?

LouisMonierSpark replied:
Yes, pas nécessaire de faire un truc méga clean comme si des externes allaient l'utiliser. Mais commence par une doc interne dans une task Airtable, un knowledge, as you want. De manière à ce que ça aille vite si on veut se servir du repo et le manipuler.

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+init_hash.update(connection_hash)
+
+# Prepare query hash (here is an example for Meta):
+query_hash = {
```
Le _query_hash_ serait différent pour d'autres plateformes comme Tiktok ? Ou le but est d'unifier l'usage ?

simeo-potiron replied:
A voir en fonction des payload nécessaires pour les autres platformes, mais ce serait top de pouvoir unifier au maximum

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,126 @@
+
+## License
+
+This project is licensed under the GNU general public License - see the [LICENSE](LICENSE) file for details.
```
Tu as regardé ça de plus près ?

simeo-potiron replied:
Yes, j'ai pris une licence un peu générique et j'ai ajouté les références aux repos dont je me suis inspiré.
N'hésite pas si tu vois des choses à reprendre / modifier

LouisMonierSpark commented on tests/test_template.py:
```diff
@@ -0,0 +1,10 @@
+import pytest
```
Est-ce qu'on voudrait pas mettre en place des unit tests ? Lancé auto à chaque nouvelle version du package ? Qui doivent forcément être validés en amont.

simeo-potiron replied:
Yes, c'est carrément l'idée de ce fichier.
Juste j'avais pas encore de tests à implémenter pour le moment

simeo-potiron replied:
Same que précédemment: Créer une feature request pour mettre en place des tests unitaires.
Il faudrait réfléchir aux éléments que l'on voudrait tester.

LouisMonierSpark replied:
A garder en to do dans l'US pour pas perdre l'idée de cette implém plutôt v2

LouisMonierSpark commented on setup.py:
```diff
@@ -0,0 +1,51 @@
+readme_filename = os.path.join(this_dir, 'README.md')
+requirements_filename = os.path.join(this_dir, 'requirements.txt')
+
+PACKAGE_NAME = "nanga_ad_library"
```
"_" --> "-"

simeo-potiron replied:
Ok

LouisMonierSpark commented on nanga_ad_library/utils/param_checker.py:
```diff
@@ -0,0 +1,111 @@
+        if isinstance(value, list):
+            # Check that all elements are of one of the expected types
+            if accepted_values and not all([x in accepted_values for x in value]):
+                # TO UPDATE
```
Pourquoi tous les ValueError ont un commentaire #TO UPDATE ?

simeo-potiron replied:
Je voulais faire une classe d'erreur spécifique, mais pas vraiment eu le temps. 
Ce serait à reprendre dans une V2 pourquoi pas: pas nécessaire au bon fonctionnement

simeo-potiron replied:
Pareil on pourrait faire une première issue ou feature request sur le répo avec l'ajout de classes d'erreur spécifiques, pour celles qui n'ont pas encore été gérées.
Wdyt ?

LouisMonierSpark replied:
Yes, as you want. Passer par une issue pourquoi pas. 
Documente bien aussi tous les dév V2 dans l'US.

LouisMonierSpark commented on nanga_ad_library/utils/version.py:
```diff
@@ -0,0 +1,140 @@
+    return version
+
+
+def get_sdk_version():
```
On parle vraiment de SDK avec ce repo github ?

simeo-potiron replied:
Je savais pas trop comment le qualifier: pas vraiment une api non plus...
Si tu as une idée de comment le qualifier

simeo-potiron replied:
Du coup j'ai tout adapté pour que ça corresponde au naming "SDK"

LouisMonierSpark commented on nanga_ad_library/utils/version.py:
```diff
@@ -0,0 +1,140 @@
+        )
+
+
+def compare_version_to_default(required_version: str, default_version: str):
```
C'est pas auto-géré par Pypi directement les versions et compagnie ?

simeo-potiron replied:
Ce sont les versions des API des ad libraries, donc non

simeo-potiron replied:
Ok pour toi ?

LouisMonierSpark replied:
Ok for me

LouisMonierSpark commented on nanga_ad_library/utils/ad_libraries_config.py:
```diff
@@ -0,0 +1,7 @@
+"""
+
+platforms_api_versions = {
+  "Meta_GRAPH_API": "v21.0"
```
C'est quelque chose qu'on veut hardcoder dans le package ? Ou laisser libre d'accès au user ? Ici c'est plutôt option 1 mais est-ce qu'on ne voudrait pas l'option 2 ?

simeo-potiron replied:
Il s'agit uniquement d'une version par défaut (la plus récente), si le user en précise une on prend celle qu'il a précisé

simeo-potiron replied:
Ok pour toi ?

LouisMonierSpark replied:
Ok for me

LouisMonierSpark commented on examples/main.py:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    # Add your personnal tokens here (provided one is expired)
+    meta_access_token = "EAACwplIENusBO97zlPpMOZAPrhCHQfAPQLucRZBPU63zZCSeQ4HZANu3OBisZA25S3E97Q5LXhd" \
```
Warning: ne pas mettre la clef API dans un repo public mais cachée dans un dossier .env

simeo-potiron replied:
Same, pas vocation à être publié

LouisMonierSpark replied:
Tu as viré le fichier main.py du repo remote coup ?

simeo-potiron replied:
yep, il avait uniquement vocation à te faire tester en local le repo pour mieux le comprendre et pour mes tests.
Pas censé être sur le répo.

LouisMonierSpark commented on examples/main.py:
```diff
@@ -0,0 +1,77 @@
+    meta_access_token = "EAACwplIENusBO97zlPpMOZAPrhCHQfAPQLucRZBPU63zZCSeQ4HZANu3OBisZA25S3E97Q5LXhd" \
+                        "2BBQJ1RAZB0T3FDnccwBdbYJ7ujWf15IgFYGQDOgTZCSO565uJWTpiyYhs9tbvQr0j5oslNoUn82" \
+                        "oliaKpGe9Sf4M8KCc0ZCGoszuuEO6dDGr9ZAPLjCi9Wja4KlcECl0N4WGNramjuy96TriapJ8ZD"
+    meta_app_secret = None
```
Cette variable ne sert pas ? Si non, pourquoi la garder ? 
Elle n'apparait notamment pas dans le README.

LouisMonierSpark commented on examples/main.py:
```diff
@@ -0,0 +1,77 @@
+
+    print("\n~~~~ Focusing on Microsoft GB ~~~~\n")
+    new_payload = {
+        "search_page_ids": [208323729264134],
```
Il nous faudrait une liste dans la doc des différents paramètres qu'on peut utiliser dans les payload.
Et on pourrait les intégrer avec un package comme argparse. Exemple d'un repo que j'avais codé il y a fort longtemps: https://github.com/louis-monier/DeepReinforcementLearningAlgorithms/blob/master/PPO/run_experiment.py

simeo-potiron replied:
Yes indeed
Pas bloquant pour une V1 j'ai l'impression, donc pourquoi pas dans une V2 (si utilisé par d'autres gens que nous très intéressant d'avoir ça !).
Très chaud si tu as un peu d'expertise là dessus de regarder ça ensemble.

LouisMonierSpark replied:
Ouais je vois comment ça fonctionne, c'est pas très compliqué et maintenant avec GPT (ou cursor sur VScode), ça prend bcp moins longtemps.
Tu peux noter en v2 dans l'US ?

LouisMonierSpark commented on examples/main.py:
```diff
@@ -0,0 +1,77 @@
+        "search_page_ids": [208323729264134],
+        "ad_reached_countries": ["GB"],
+        "ad_status": "ACTIVE",
+        "fields": ["ad_delivery_start_time", "ad_delivery_stop_time"]
```
Idem, ici il faudrait qu'on puisse rechercher facilement ce qu'on peut inclure comme noms de fields

simeo-potiron replied:
Pareil que plus haut

LouisMonierSpark commented on examples/main.py:
```diff
@@ -0,0 +1,77 @@
+from nanga_ad_library import NangaAdLibraryApi
+
+
+def iter_on_results(results, limit):
```
Est-ce qu'on ne mettrait pas cette fonction dans utils si elle sert pour récupérer les résultats ?

simeo-potiron replied:
Ce fichier était vraiment là pour te permettre de comprendre comment marchait le module. Il n'a pas vocation a être public.

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+        self._version = get_default_api_version(self.API_NAME)
+
+        # Other useful components
+        self._target_political_ads = False
```
Toujours False ? Pas un param qu'on veut potentiellement rendre dispo ?

LouisMonierSpark replied:
Un avis là-dessus ?

simeo-potiron replied:
Cf ligne 118/131 (dans le init). 
J'ai besoin de le modifier après avoir initialisé l'objet, d'où le = False
Je vais ajouter un comment

LouisMonierSpark replied:
Ok ça marche. Je laisse la conversation ouverte pour avoir le rappel.

simeo-potiron replied:
Obligé de la clore pour merge la PR du coup

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+
+class AdCountry(Enum):
+    """
+        Values available for ad_reached_countries parameter.
```
Petit tab en trop

simeo-potiron replied:
Yes j'ai enlevé merci, bien vu !

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+
+class AdType(Enum):
+    """
+        Values available for ad_type parameter.
```
Idem et partout dans la suite.

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+    SWEDEN = "SE"
+
+
+class MetaParam(Enum):
```
Ah stylé la gestion des params

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+    def get_method(self):
+        return self._method
+
+    def update_method(self, method: str):
```
Est-ce qu'on veut vraiment ça ?

LouisMonierSpark replied:
Je veux dire, est-ce qu'un update est nécessaire ici ? 
On veut juste du get.

simeo-potiron replied:
Oui c'est vrai à priori pour le moment, même si peut dépendre des API (post peut renvoyer des infos aussi).
J'ai essayé de faire quelque chose de générique: get_method & update_methode permettent de récupérer et de modifier des attributs cachés de cette classe.
Ok pour toi si on laisse comme ça ?

LouisMonierSpark replied:
Ok for me

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1287 @@
+
+
+"""
+Useful classes to help MetaAdLibrary objects creation.
```
Tout dans le même fichier ? On pourrait ventiler dans des fichiers type "utils"

simeo-potiron replied:
C'est vraiment relatif à cette classe, plutôt chaud de garder ça là personnellement.
Ok pour toi ?

simeo-potiron replied:
J'ai répondu à tout normalement now

LouisMonierSpark commented on nanga_ad_library/api.py:
```diff
@@ -0,0 +1,408 @@
+
+
+"""
+Some useful functions.
```
J'ai la sensation que tout le utils doit être rangé d'un côté pour se concentrer que sur ce qui nous intéresse ici: la requête à l'api

LouisMonierSpark replied:
T'en penses quoi ?

simeo-potiron replied:
Yes alright, j'ai créé trois nouveaux fichiers utils:
- object_parser: 
  - avec la classe ObjectParser
- request_handler: 
  - avec les classes PlatformResponse et HttpMethod
  - avec la fonction json_encode_top_level_param.

La classe ResultCursor me parait trop liée à ? NangaAdLibrary pour le déplacer dans utils.

LouisMonierSpark replied:
Thanks ! Bien clean comme ça ! 
Ok for me

LouisMonierSpark commented on nanga_ad_library/api.py:
```diff
@@ -0,0 +1,408 @@
+"""
+
+
+def json_encode_top_level_param(params):
```
Pourquoi pas formatté correctement avant ?

simeo-potiron replied:
Ici pour le coup c'est une fonction 100% tirée du repo de Meta: c'est pour permettre de passer des dicts (coté user) et de l'envoyer sous un format approprié (str) à l'API Meta.

LouisMonierSpark replied:
Ok ça marche

LouisMonierSpark commented on nanga_ad_library/exceptions.py:
```diff
@@ -0,0 +1,121 @@
+    pass
+
+
+class PlatformRequestError(PlatformError):
```
Ah nice comme classe

simeo-potiron replied:
Fortement inspiré de la classe d'erreur du répo de Meta pour le coup haha

LouisMonierSpark commented on nanga_ad_library/sessions.py:
```diff
@@ -0,0 +1,311 @@
+"""
+
+
+class ApiSession(object):
```
A quel point tout ça est utile ? N'est-ce pas encapsulé dans le package requests ? A quel point sera utile dans le day 2 day ?

simeo-potiron replied:
Ca permet de faire plusieurs requêtes, de mettre à jour les payload etc...
Ta session est encapsulée dans ton API

LouisMonierSpark reviewed (commented):
Solide repo ! Congrats ! Je ne connais pas le % de ré-utilisation d'autres repos mais il y a une gross quantité de travail. Quelques remarques générales avant de voir celles dans le code lui-même:
- Checker la LICENCE. Je ne sais pas comment ça fonctionne concrètement.
- Supprimer tout le dossier et fichiers **.idea**
- Supprimer tout le dossier et fichiers **utils/_pycache_**
- Supprimer tout le dossier et fichiers **nanga_ad_library/ad_libraries/__pycache__**
- Pourquoi tous les ValueError ont un commentaire "To update" ?
- Attention à la clef API hardcodée dans un des fichiers py

simeo-potiron committed:
Quelques corrections

simeo-potiron committed:
Ignore "examples" folder

simeo-potiron reviewed (commented):
Réponse à tes commentaires

simeo-potiron committed:
quick fix

simeo-potiron committed:
Add ad downloader

- Add ad downloader to repo
- Protect hidden methods and attributes ("_hidden" => "__hidden")
- Use sdk to describe the tool
- Remove unwanted files

LouisMonierSpark commented on README.md:
```diff
@@ -1,2 +1,125 @@
+
+## Acknowledgements
+
+- [Meta Business SDK](https://github.com/facebook/facebook-python-business-sdk) was a great inspiration for this work. 
```
Mettre aussi l'inspi du repo pour l'ad downloader

simeo-potiron replied:
Pas d'inspi pour l'ad_downloader, ça vient de chez moi ;)

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+
+class MetaRequestInterceptor:
+    """
+    A class used to intercept requests from Meta Ad Library preview while downloading ad elements.
```
A class designed to passively intercept network requests from the Meta Ad Library preview while downloading ad elements.
It blocks direct video requests (to avoid detection) and captures video and image URLs instead. These URLs are stored for extracting video links and thumbnails (in self.__intercepted_videos & self.__intercepted_images), enabling a bypass of Meta's anti-scraping measures.

LouisMonierSpark replied:
"ad elements." --> quelques noms des choses que l'on peut récupérer seraient les bienvenus.

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+    """
+
+    def __init__(self, verbose):
+        self.__verbose = verbose
```
C'est voulu d'utiliser du _name mangling_ ici pour le nom des variables ?

simeo-potiron replied:
yes

simeo-potiron replied:
Ce sont des éléments internes, on ne veut pas qu'ils soient accessibles de l'extérieur.

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+
+class MetaAdDownloader:
+    """
+    A class instancing a scrapper to retrieve elements from Meta Ad Library previews.
```
Un seul "p" à scraper/scraping.

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+    """
+    A class instancing a scrapper to retrieve elements from Meta Ad Library previews.
+
+    Adds extracted elements to each ad from a batch using the preview.
```
Adds extracted elements from the preview to each ad from a batch.

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+    Adds extracted elements to each ad from a batch using the preview.
+    """
+
+    PREVIEW_FIELD = "ad_snapshot_url"
```
Qu'est-ce que c'est ? Explication needed !

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+    PREVIEW_FIELD = "ad_snapshot_url"
+
+    def __init__(self, verbose=False):
+        self.__verbose = verbose or False
```
Redondant si argument avec valeur par défaut définie comme False.

simeo-potiron replied:
Pas forcément: si verbose n'est pas précisé, c'est la valeur par défaut qui est utilisée (verbose=False).
Si verbose est précisé mais est None, c'est la valeur False qui est appliquée via verbose or False.
Aligné ?

LouisMonierSpark replied:
Indeed
Ok for me

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+        Install Playwright chromium browser (async)
+
+        Args:
+            with_deps (bool, optional): Wether deps have to be installed too.
```
"whether" sinon ça veut dire "castrated male sheep" 😄

simeo-potiron replied:
Il s'agit d'une coquille, je voulais évidemment utiliser "Whether"  😅

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+            args.append("--with-deps")
+
+        # Launch install cmd
+        proc = subprocess.run(args, env=get_driver_env(), capture_output=True, text=True, check=False)  # noqa: S603
```
# noqa: S603 ?

simeo-potiron replied:
Exécuter une ligne de commande shell dans du code n'est pas un comportement recommandé, mais cet indicatif permet de signaler que son utilisation ici est volontaire et assumée. 
cf [doc](https://flake8.pycqa.org/en/latest/user/violations.html#selecting-and-ignoring-violations) (ici violation flake8 Bandit S603 sur l'exécution de lignes de commandes avec subprocess).
C'était dans l'exemple que j'ai trouvé, j'ai pensé que ça pouvait être plus clean même si pas de vérification de la sécurité du code chez nous en pratique.

LouisMonierSpark replied:
Ah ok nice j'avais jamais vu ça
Interesting

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+
+    async def download_from_new_batch(self, ad_library_batch):
+        """
+        Use parallelized calls to download ad elements for each row of a batch
```
Donc 1 batch = 1 row dans l'ad library = 3 encarts ?

simeo-potiron replied:
Non 1 batch = 25 lignes de l'ad library, ça correspond aux batchs du curseur (cf class ResultCursor du fichier sdk.py).

LouisMonierSpark replied:
Ok for me

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+        """
+        Use parallelized calls to download ad elements for each row of a batch
+
+        :param ad_library_batch: A list of records from
```
from ... ?

simeo-potiron replied:
Petit oubli

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+        return updated_batch
+
+    async def __download_ad_elements(self, browser, ad_payload):
+        """ [Hidden method]
```
Comme les autres avec du nom manglée. A ajouter à __install_chromium_browser_async & __gen_app_secret_proof par consistence dans les comments.

simeo-potiron replied:
Yes, indeed

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+
+    async def __download_ad_elements(self, browser, ad_payload):
+        """ [Hidden method]
+        Use scrapping to extract all ad elements from the ad preview url.
```
1 p

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+        await page.route("**/*", interceptor.intercept)
+
+        # Open Ad Library card and wait until all requests are finished / Increase nav timeout to 60s
+        await page.goto(preview, timeout=300000)
```
5 minutes plutôt non ?

simeo-potiron replied:
Indeed, oublié d'adapter le comment après avoir augmenté.

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+                ad_elements["carousel"].append(creative)
+
+        # Deal with ads displaying only one creative
+        else:
```
La logique est la même à de nombreux endroits: un rassemblement pourrait être envisageable.

simeo-potiron replied:
J'ai rassemblé au max que je pouvais: selon le fait que l'on soit en carousel ou pas les XPath sont sensiblement différents.
Si tu vois quelque chose de plus efficace/clean, chaud d'avoir ton input, je suis peu être passé à coté de quelque chose.

LouisMonierSpark replied:
J'ai pas regardé exactement en détail mais la structure apparaissait très similaire. 
Je te fais confiance

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+            # Add to list
+            ad_elements["carousel"].append(creative)
+
+        # time.sleep(60)
```
No need for sleep ?

simeo-potiron replied:
Nope, c'était pour mes tests (en headless = False). 
Je kill.
Bien vu !

LouisMonierSpark commented on nanga_ad_library/ad_downloaders/meta_ad_downloader.py:
```diff
@@ -0,0 +1,312 @@
+            browser = await p.chromium.launch()
+
+            # Download ad elements (parallel calls)
+            updated_batch = await asyncio.gather(
```
Combien de lancements en // ? Je ne vois pas d'indications sur une potentielle restriction.

LouisMonierSpark replied:
Comment sont formés ces batchs ? Combien d'éléments contiennent-ils ? 3 ?
Combien de batchs sont lancés en même temps ?
Ca ne m'apparait pas clairement à la lecture du code. As-tu plus d'indications dans une task Airtable ou autre ? Il faut documenter un peu plus.

simeo-potiron replied:
Pas de restrictions, à priori autant que l'on veut.
Ici en réalité on est à 25 (sur Meta), et ils sont en fait lié à l'utilisation du ResponseCursor.

LouisMonierSpark replied:
Pas besoin de mettre des garde fous donc ?

simeo-potiron replied:
A priori non, il y a le timeout qui crashera si jamais l'usage est abusif

LouisMonierSpark commented on nanga_ad_library/ad_libraries/meta_ad_library.py:
```diff
@@ -0,0 +1,1280 @@
+import warnings
```
Il y a ey des modifications apparemment sur ce fichier mais pas de rép aux comments.

simeo-potiron replied:
Ok, je regarde

LouisMonierSpark reviewed (commented):
Pas mal de questions/remarques sur _meta_ad_downloader_.
Ca manque de documentation. En quelques phrases, je pense que la logique devient naturelle à comprendre plutôt que de devoir la découvrir en lisant le code.
Remarque: tu as vu que tu as mis sparkgeeks en reviewer ?

simeo-potiron reviewed (commented):
J'ai répondu à tes commentaires.

J'en ai profité pour harmoniser un peu mes plages de comments des fontions:
1: 
```
""" 
:param: arg: arg description
:return: returned object description
:member: ...
"""
```
vs
2: 
```
"""
Args:
    arg1: arg1 description
    arg2: arg2 description

Returns:
    Returned object desciption

Raises:
...
"""
```
J'ai adopté le template 2 partout: j'avais un peu des 2 selon les endroits.

simeo-potiron committed:
Fix PR comments

Update code using Louis propositions (cf PR)
+ Harmonize comment section templates

simeo-potiron committed:
Add dowloader to README file

simeo-potiron committed:
Last adjustments

Move some classes and functions in utils.
+ small updates

simeo-potiron reviewed (commented):
Sorry j'étais passé à coté de certains comments qui étaient sur des fichiers supp ou renommés.
J'ai répondu à tout et fait quelques ajustement dans mon dernier commit grâce à tes recommendations.

simeo-potiron committed:
Reshape extracted landing page
